File: bench/bench_merge.py
import random
import gc
import time
from pandas import *
from pandas.compat import range, lrange, StringIO
from pandas.util.testing import rands

N = 10000
ngroups = 10


def get_test_data(ngroups=100, n=N):
    unique_groups = lrange(ngroups)
    arr = np.asarray(np.tile(unique_groups, n / ngroups), dtype=object)

    if len(arr) < n:
        arr = np.asarray(list(arr) + unique_groups[:n - len(arr)],
                         dtype=object)

    random.shuffle(arr)
    return arr

# aggregate multiple columns
# df = DataFrame({'key1' : get_test_data(ngroups=ngroups),
#                 'key2' : get_test_data(ngroups=ngroups),
#                 'data1' : np.random.randn(N),
#                 'data2' : np.random.randn(N)})

# df2 = DataFrame({'key1'  : get_test_data(ngroups=ngroups, n=N//10),
#                  'key2'  : get_test_data(ngroups=ngroups//2, n=N//10),
#                  'value' : np.random.randn(N // 10)})
# result = merge.merge(df, df2, on='key2')

N = 10000

indices = np.array([rands(10) for _ in range(N)], dtype='O')
indices2 = np.array([rands(10) for _ in range(N)], dtype='O')
key = np.tile(indices[:8000], 10)
key2 = np.tile(indices2[:8000], 10)

left = DataFrame({'key': key, 'key2': key2,
                  'value': np.random.randn(80000)})
right = DataFrame({'key': indices[2000:], 'key2': indices2[2000:],
                   'value2': np.random.randn(8000)})

right2 = right.append(right, ignore_index=True)


join_methods = ['inner', 'outer', 'left', 'right']
results = DataFrame(index=join_methods, columns=[False, True])
niter = 10
for sort in [False, True]:
    for join_method in join_methods:
        f = lambda: merge(left, right, how=join_method, sort=sort)
        gc.disable()
        start = time.time()
        for _ in range(niter):
            f()
        elapsed = (time.time() - start) / niter
        gc.enable()
        results[sort][join_method] = elapsed
# results.columns = ['pandas']
results.columns = ['dont_sort', 'sort']


# R results
# many to one
r_results = read_table(StringIO("""      base::merge   plyr data.table
inner      0.2475 0.1183     0.1100
outer      0.4213 0.1916     0.2090
left       0.2998 0.1188     0.0572
right      0.3102 0.0536     0.0376
"""), sep='\s+')

presults = results[['dont_sort']].rename(columns={'dont_sort': 'pandas'})
all_results = presults.join(r_results)

all_results = all_results.div(all_results['pandas'], axis=0)

all_results = all_results.ix[:, ['pandas', 'data.table', 'plyr',
                                 'base::merge']]

sort_results = DataFrame.from_items([('pandas', results['sort']),
                                     ('R', r_results['base::merge'])])
sort_results['Ratio'] = sort_results['R'] / sort_results['pandas']


nosort_results = DataFrame.from_items([('pandas', results['dont_sort']),
                                       ('R', r_results['base::merge'])])
nosort_results['Ratio'] = nosort_results['R'] / nosort_results['pandas']

# many to many

# many to one
r_results = read_table(StringIO("""base::merge   plyr data.table
inner      0.4610 0.1276     0.1269
outer      0.9195 0.1881     0.2725
left       0.6559 0.1257     0.0678
right      0.6425 0.0522     0.0428
"""), sep='\s+')

all_results = presults.join(r_results)
all_results = all_results.div(all_results['pandas'], axis=0)
all_results = all_results.ix[:, ['pandas', 'data.table', 'plyr',
                                 'base::merge']]

File: bench/bench_merge.R
library(plyr)
library(data.table)
N <- 10000
indices = rep(NA, N)
indices2 = rep(NA, N)
for (i in 1:N) {
  indices[i] <- paste(sample(letters, 10), collapse="")
  indices2[i] <- paste(sample(letters, 10), collapse="")
}
left <- data.frame(key=rep(indices[1:8000], 10),
                   key2=rep(indices2[1:8000], 10),
                   value=rnorm(80000))
right <- data.frame(key=indices[2001:10000],
                    key2=indices2[2001:10000],
                    value2=rnorm(8000))

right2 <- data.frame(key=rep(right$key, 2),
                     key2=rep(right$key2, 2),
                     value2=rnorm(16000))

left.dt <- data.table(left, key=c("key", "key2"))
right.dt <- data.table(right, key=c("key", "key2"))
right2.dt <- data.table(right2, key=c("key", "key2"))

# left.dt2 <- data.table(left)
# right.dt2 <- data.table(right)

## left <- data.frame(key=rep(indices[1:1000], 10),
##                    key2=rep(indices2[1:1000], 10),
##                    value=rnorm(100000))
## right <- data.frame(key=indices[1:1000],
##                     key2=indices2[1:1000],
##                     value2=rnorm(10000))

timeit <- function(func, niter=10) {
  timing = rep(NA, niter)
  for (i in 1:niter) {
    gc()
    timing[i] <- system.time(func())[3]
  }
  mean(timing)
}

left.join <- function(sort=FALSE) {
  result <- base::merge(left, right, all.x=TRUE, sort=sort)
}

right.join <- function(sort=FALSE) {
  result <- base::merge(left, right, all.y=TRUE, sort=sort)
}

outer.join <- function(sort=FALSE) {
  result <- base::merge(left, right, all=TRUE, sort=sort)
}

inner.join <- function(sort=FALSE) {
  result <- base::merge(left, right, all=FALSE, sort=sort)
}

left.join.dt <- function(sort=FALSE) {
  result <- right.dt[left.dt]
}

right.join.dt <- function(sort=FALSE) {
  result <- left.dt[right.dt]
}

outer.join.dt <- function(sort=FALSE) {
  result <- merge(left.dt, right.dt, all=TRUE, sort=sort)
}

inner.join.dt <- function(sort=FALSE) {
  result <- merge(left.dt, right.dt, all=FALSE, sort=sort)
}

plyr.join <- function(type) {
  result <- plyr::join(left, right, by=c("key", "key2"),
                       type=type, match="first")
}

sort.options <- c(FALSE, TRUE)

# many-to-one

results <- matrix(nrow=4, ncol=3)
colnames(results) <- c("base::merge", "plyr", "data.table")
rownames(results) <- c("inner", "outer", "left", "right")

base.functions <- c(inner.join, outer.join, left.join, right.join)
plyr.functions <- c(function() plyr.join("inner"),
                    function() plyr.join("full"),
                    function() plyr.join("left"),
					function() plyr.join("right"))
dt.functions <- c(inner.join.dt, outer.join.dt, left.join.dt, right.join.dt)
for (i in 1:4) {
  base.func <- base.functions[[i]]
  plyr.func <- plyr.functions[[i]]
  dt.func <- dt.functions[[i]]
  results[i, 1] <- timeit(base.func)
  results[i, 2] <- timeit(plyr.func)
  results[i, 3] <- timeit(dt.func)
}


# many-to-many

left.join <- function(sort=FALSE) {
  result <- base::merge(left, right2, all.x=TRUE, sort=sort)
}

right.join <- function(sort=FALSE) {
  result <- base::merge(left, right2, all.y=TRUE, sort=sort)
}

outer.join <- function(sort=FALSE) {
  result <- base::merge(left, right2, all=TRUE, sort=sort)
}

inner.join <- function(sort=FALSE) {
  result <- base::merge(left, right2, all=FALSE, sort=sort)
}

left.join.dt <- function(sort=FALSE) {
  result <- right2.dt[left.dt]
}

right.join.dt <- function(sort=FALSE) {
  result <- left.dt[right2.dt]
}

outer.join.dt <- function(sort=FALSE) {
  result <- merge(left.dt, right2.dt, all=TRUE, sort=sort)
}

inner.join.dt <- function(sort=FALSE) {
  result <- merge(left.dt, right2.dt, all=FALSE, sort=sort)
}

sort.options <- c(FALSE, TRUE)

# many-to-one

results <- matrix(nrow=4, ncol=3)
colnames(results) <- c("base::merge", "plyr", "data.table")
rownames(results) <- c("inner", "outer", "left", "right")

base.functions <- c(inner.join, outer.join, left.join, right.join)
plyr.functions <- c(function() plyr.join("inner"),
                    function() plyr.join("full"),
                    function() plyr.join("left"),
					function() plyr.join("right"))
dt.functions <- c(inner.join.dt, outer.join.dt, left.join.dt, right.join.dt)
for (i in 1:4) {
  base.func <- base.functions[[i]]
  plyr.func <- plyr.functions[[i]]
  dt.func <- dt.functions[[i]]
  results[i, 1] <- timeit(base.func)
  results[i, 2] <- timeit(plyr.func)
  results[i, 3] <- timeit(dt.func)
}

